IBX-10684: Document translation management (#3249)
ezsystems/developer-documentation: docs/api/event_reference/translations_management_events.md

---
description: Events that are triggered when working with translations management.
edition: lts-update
page_type: reference
---

# Translations management events

The [Translations management](configure_translations_management.md) package dispatches events at two levels.

## Translation events

> Translation events are thrown once per field value per translation operation.

Reviewer: ```suggestion
Translation events are dispatched once per field value per translation operation.
```

Exceptions are thrown (`throw new InvalidArgumentException`), event are dispatched by the event dispatcher ( https://github.com/ibexa/translations-management/blob/9e62c293b8085443c16f10a34de3b089f79b126b/src/lib/AutoTranslate/Event/EventDispatchingProviderTranslator.php#L45 )

I'm not sure I understand this sentence correctly - for each field type to be translated and for each language that field is translated into?

3 fields, 2 languages - 6 before and after events? 

How about:

``` suggestion
Translation events are dispatched for every field in each selected target language.
```

> They are used for logging, analytics, and observability.

Reviewer: Are they used, or can they be used?

I think the system uses them only for logging by default (https://github.com/ibexa/translations-management/blob/9e62c293b8085443c16f10a34de3b089f79b126b/src/bundle/AutoTranslate/EventSubscriber/TranslationEventSubscriber.php#L11 ? )

```suggestion
Use them for logging, analytics, and observability.
```

Both events are read-only, you can't use them to override the translation result.

| Event | Dispatched by | Dispatched when | Properties |
|---|---|---|----|

> | [`BeforeTranslateEvent`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-TranslationsManagement-AutoTranslate-Event-BeforeTranslateEvent.html) | `TranslationService` | Before a translation request is sent to the provider | `TranslationProviderInterface $provider`</br>`string $text`</br>`string $sourceLanguage`</br>`string $targetLanguage` |

Reviewer 2: ```suggestion
| [`BeforeTranslateEvent`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-TranslationsManagement-AutoTranslate-Event-BeforeTranslateEvent.html) | `EventDispatchingProviderTranslator` | Before a translation request is sent to the provider | `TranslationProviderInterface $provider`</br>`string $text`</br>`string $sourceLanguage`</br>`string $targetLanguage` |
```

> | [`BeforeTranslateEvent`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-TranslationsManagement-AutoTranslate-Event-BeforeTranslateEvent.html) | `EventDispatchingProviderTranslator` | Before a translation request is sent to the provider | `TranslationProviderInterface $provider`</br>`string $text`</br>`string $sourceLanguage`</br>`string $targetLanguage` |

Reviewer: Please replace all `</br>` usages with `<br>`.

`</br>` is not valid HTML and causes issues when converting the result to Markdown for AI Agents (https://github.com/ibexa/documentation-developer/pull/3297#discussion_r3621656184)

BTW, I don't think we should document the properties - with the PHP API Reference available, that information is there, generated for us automatically, and we don't have to maintain it.

| [`TranslateEvent`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-TranslationsManagement-AutoTranslate-Event-TranslateEvent.html) | `EventDispatchingProviderTranslator` | After a translation response is received |

## Side-by-side creation events

> Side-by-side creation events are dispatched when a new translation draft is being prepared.

Reviewer 3: ```suggestion
Side-by-side creation events are dispatched when preparing a new translation draft.
```

> | [`TranslateEvent`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-TranslationsManagement-AutoTranslate-Event-TranslateEvent.html) | `TranslationService` | After a translation response is received | `string $result`</br>`TranslationProviderInterface $provider`</br>`string $text`</br>`string $sourceLanguage`</br>`string $targetLanguage` |

Reviewer 2: ```suggestion
| [`TranslateEvent`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-TranslationsManagement-AutoTranslate-Event-TranslateEvent.html) | `EventDispatchingProviderTranslator` | After a translation response is received | `string $result`</br>`TranslationProviderInterface $provider`</br>`string $text`</br>`string $sourceLanguage`</br>`string $targetLanguage` |
```

| Event | Dispatched by | Dispatched when |
|---|---|---|

> | [`OnContentSideBySideTranslationCreateEvent`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-TranslationsManagement-SideBySide-Event-OnContentSideBySideTranslationCreateEvent.html) | `ContentTranslationCreateController` | When a draft side-by-side translation of a content item is being created |

Reviewer 3: Same as above - I'd rewrite it to avoid passive tense.

> | [`OnProductSideBySideTranslationCreateEvent`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-TranslationsManagement-SideBySide-Event-OnProductSideBySideTranslationCreateEvent.html) | `ProductTranslationViewController` | When a draft side-by-side translation of a product is being created |

Reviewer 3: Same.

> | [`OnContentSideBySideTranslationCreateEvent`](/api/php_api/php_api_reference/classes/Ibexa-Contracts-TranslationsManagement-SideBySide-Event-OnContentSideBySideTranslationCreateEvent.html) | `SideBySideTranslationService` | When a draft side-by-side translation of a content item is being created | `Request $request`</br>`Content $sourceContent`</br>`string $sourceLanguageCode`</br>`string $targetLanguageCode`</br>`?Content $targetDraft` |

Reviewer 4: The "Dispatched by" column is inaccurate for the side-by-side events. In the current implementation, `OnContentSideBySideTranslationCreateEvent` is dispatched from `ContentTranslationCreateController`, and `OnProductSideBySideTranslationCreateEvent` is dispatched from `ProductTranslationViewController`, not from `SideBySideTranslationService`. Could we update this so the reference matches the actual emitter?
